### PR TITLE
Feature/sdl 0192 button subscription response from hmi [internal review]

### PR DESCRIPTION
**Pull request by VjKlepikov**
ATF Test Scripts to check [FORDTCN-12168](https://adc.luxoft.com/jira/browse/FORDTCN-12168)

This PR is **[ready]** for review.

### Summary
Scripts for "Button subscription response from hmi"

### ATF version
[8.0.0_version_bump](https://github.com/smartdevicelink/sdl_atf/tree/feature/8.0.0_version_bump)

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: test_scripts/API/ButtonSubscription/002_Subscribe_CUSTOM_BUTTON_Success.lua [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+local common = require('test_scripts/API/ButtonSubscription/common_buttonSubscription')
+
+--[[ Local Variables ]]
+local appSessionId1 = 1
+local buttonName = "CUSTOM_BUTTON"
+
+--[[ Scenario ]]
+common.runner.Title("Preconditions")
+common.runner.Step("Clean environment", common.preconditions)
+common.runner.Step("Start SDL, HMI, connect Mobile, start Session", common.start)
```
Comment (VNikiforov-dev) [inline test_scripts/API/ButtonSubscription/002_Subscribe_CUSTOM_BUTTON_Success.lua:25]: add button to HMI capabilities during start HMI:
common.runner.Step("Start SDL, HMI, connect Mobile, start Session", common.start, { common.addButtonToCapabilities(common.custom_button_obj) })

Comment (VNikiforov-dev) [reply test_scripts/API/ButtonSubscription/002_Subscribe_CUSTOM_BUTTON_Success.lua:25]: addButtonToCapabilities implemented in common_buttonSubscription.lua in PR https://github.com/LuxoftSDL/sdl_atf_test_scripts/pull/127

File: test_scripts/API/ButtonSubscription/002_Subscribe_CUSTOM_BUTTON_Success.lua [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+common.runner.Title("Test")
+common.runner.Step("App registration and check Subscribe on CUSTOM_BUTTON", common.registerAppSubCustomButton)
+common.runner.Step("Activate app", common.activateApp)
+common.runner.Step("Subscribe on Soft button", common.registerSoftButton)
```
Comment (VNikiforov-dev) [inline test_scripts/API/ButtonSubscription/002_Subscribe_CUSTOM_BUTTON_Success.lua:30]: Extra step in the test case?

File: test_scripts/API/ButtonSubscription/005_Subscribe_button_SDL_send_UnsubscribeButton_after_Subscribe_SUCCESS_timeout.lua [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+-- 2) Receive response Buttons.UnsubscribeButton(SUCCESS) and keep actual unsubscribed state for <button>
+-- 3) Not send UnsubscribeButton response to mobile app
+------------------------------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/ButtonSubscription/common_buttonSubscription')
+local events = require('events')
+
+--[[ Local Variables ]]
+local appSessionId1 = 1
+local himCID
```
Comment (VNikiforov-dev) [inline test_scripts/API/ButtonSubscription/005_Subscribe_button_SDL_send_UnsubscribeButton_after_Subscribe_SUCCESS_timeout.lua:25]: himCID is used only in a function rpcGenericError. Maybe moving the variable declaration to function? (to narrow the scope)

File: test_scripts/API/ButtonSubscription/common_buttonSubscription.lua [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+--! @parameters:
+--! pAppId - application number (1, 2, etc.)
+--! pExpTimesApp - number of notifications for 1st app
+--! @return: none
+--]]
+function m.registerAppSubCustomButton(pAppId, pExpTimesApp)
+  if not pAppId then pAppId = 1 end
+  if not pExpTimesApp then pExpTimesApp = m.isExpected end
+  m.registerAppWOPTU(pAppId)
+  m.getHMIConnection(pAppId):ExpectRequest("Buttons.SubscribeButton", { m.getHMIAppId(), buttonName = "CUSTOM_BUTTON" })
```
Comment (VNikiforov-dev) [inline test_scripts/API/ButtonSubscription/common_buttonSubscription.lua:380]:
```suggestion
  m.getHMIConnection(pAppId):ExpectRequest("Buttons.SubscribeButton", { appID = m.getHMIAppId(pAppId), buttonName = "CUSTOM_BUTTON" })
```

File: test_scripts/SDL5_0/MobileVersioning/PlayPauseButton/002_play_pause_v4_invalid_data.lua [hunk 5]
```diff
@@ -42,7 +42,11 @@ config.application1.registerAppInterfaceParams.syncMsgVersion = {
 local function subscribeButtonSuccess(pButName)
   local cid = commonSmoke.getMobileSession():SendRPC("SubscribeButton", { buttonName = pButName })
   local appIDvalue = commonSmoke.getHMIAppId()
-  EXPECT_HMINOTIFICATION("Buttons.OnButtonSubscription", { appID = appIDvalue, name = "PLAY_PAUSE", isSubscribed = true })
+   commonSmoke.getHMIConnection():ExpectRequest("Buttons.SubscribeButton",
+    { appID = appIDvalue, buttonName = "PLAY_PAUSE" })
+  :Do(function(_, data)
+      commonSmoke.getHMIConnection():SendResponse(data.id, data.method, "SUCCESS", { })
+    end)
```
Comment (dboltovskyi) [inline test_scripts/SDL5_0/MobileVersioning/PlayPauseButton/002_play_pause_v4_invalid_data.lua:49]: In line `56` there is an expectation on `Buttons.OnButtonSubscription`. Suppose it needs to be replaced by `Buttons.UnsubscribeButton` request

**Review by dboltovskyi — CHANGES_REQUESTED** [review]
@VjKlepikov Please notice:

This is 1st part of the review which is related to existing files/scripts.
2nd part will be related to a new scripts.

Besides the files in this PR please make the following updates:

1. Remove expectation on `Buttons.OnButtonSubscription` in all files in `user_modules` folder and sub-folders (or replace by appropriate new request where applicable)

2. Prepare a separate PR with removal of `Buttons.OnButtonSubscription` in `sdl_atf` repo in `modules/connecttest.lua`

File: test_scripts/Smoke/API/042_SubscribeButton_Non_Media_PositiveCase_SUCCESS.lua [hunk 6]
```diff
@@ -46,30 +54,25 @@ local buttonName = {
...
-  "PRESET_8"
-}
-
-local mediaButtonName = {
-  "PLAY_PAUSE",
-  "SEEKLEFT",
-  "SEEKRIGHT",
-  "TUNEUP",
-  "TUNEDOWN"
+  "PRESET_8",
```
[the remark this answers is not in the document]
Comment (VjKlepikov) [reply test_scripts/Smoke/API/042_SubscribeButton_Non_Media_PositiveCase_SUCCESS.lua:57]: according to requirements https://adc.luxoft.com/jira/browse/APPLINK-14516?src=confmacro

File: test_scripts/iAP2TransportSwitch/001_happy_path.lua [hunk 7]
```diff
@@ -131,7 +131,7 @@ runner.Step("Start SDL, HMI", common.start)
 
 runner.Title("Test")
 runner.Step("Connect Bluetooth Device", connectBluetoothDevice)
-runner.Step("Connect USB Device", connectUSBDevice)
+--runner.Step("Connect USB Device", connectUSBDevice)
```
[the remark this answers is not in the document]
Comment (VjKlepikov) [reply test_scripts/iAP2TransportSwitch/001_happy_path.lua:134]: my mistake during debug

**Comment by VjKlepikov** [comment]
@dboltovskyi All the above comments have been processed
PR for ATF is created https://github.com/LuxoftSDL/sdl_atf/pull/7

File: test_scripts/API/ButtonSubscription/038_Subscribe_CUSTOM_BUTTON_HMI_does_not_respond.lua [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+local errorCode = "GENERIC_ERROR"
+
+--[[ Local Functions ]]
+local function registerAppSubCustomButton(pAppId, pExpTimesApp)
+  if not pAppId then pAppId = 1 end
+  if not pExpTimesApp then pExpTimesApp = common.isExpected end
+  common.registerAppWOPTU(pAppId)
+  common.getHMIConnection(pAppId):ExpectRequest("Buttons.SubscribeButton",
+    { common.getHMIAppId(), buttonName = "CUSTOM_BUTTON" })
+ :Do(function()
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/038_Subscribe_CUSTOM_BUTTON_HMI_does_not_respond.lua:30]: Please fix indentation

File: test_scripts/API/ButtonSubscription/040_Subscribe_CUSTOM_BUTTON_twise_IGNORED.lua [hunk 9]
```diff
@@ -0,0 +1,40 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/040_Subscribe_CUSTOM_BUTTON_twise_IGNORED.lua:1]: Typo in script name: `twise`

**Review by dboltovskyi — CHANGES_REQUESTED** [review]
Please notice this is 2nd iteration of review.

File: test_scripts/API/ButtonSubscription/009_SubscribeButton_INVALID_request_INVALID_DATA.lua [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  local cid = common.getMobileSession().correlationId + 40
+  local msg = {
+    serviceType      = 7,
+    frameInfo        = 0,
+    rpcType          = 0,
+    rpcFunctionId    = 18,
+    rpcCorrelationId = cid,
+    payload          = '{"buttonName":"' .. pButtonName .. '", {}}'
+  }
+  common.getMobileSession():Send(msg)
```
[the remark this answers is not in the document]
Comment (VjKlepikov) [reply test_scripts/API/ButtonSubscription/009_SubscribeButton_INVALID_request_INVALID_DATA.lua:39]: The basic idea to check for incorrect structure ({... , {}}) I'm not sure if we can check this with SendRPC ()

File: test_scripts/API/ButtonSubscription/011_SubscribeButton_twise_the_same_button_IGNORED.lua [hunk 11]
```diff
@@ -0,0 +1,44 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/011_SubscribeButton_twise_the_same_button_IGNORED.lua:1]: Typo in script name: `twise`.

File: test_scripts/API/ButtonSubscription/012_UnsunscribeButton_twise_the_same_button_IGNORED.lua [hunk 12]
```diff
@@ -0,0 +1,46 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/012_UnsunscribeButton_twise_the_same_button_IGNORED.lua:1]: Typo in script name: `twise`.

File: test_scripts/API/ButtonSubscription/027_SubscribeButton_Resumption_Unexpected_disconect.lua [hunk 13]
```diff
@@ -0,0 +1,52 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/027_SubscribeButton_Resumption_Unexpected_disconect.lua:1]: Typo in script name: `disconect`.

File: test_scripts/API/ButtonSubscription/029_SubscribeButton_Unexpected_disconect_HMI_responds_ERROR_CODE.lua [hunk 14]
```diff
@@ -0,0 +1,65 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/029_SubscribeButton_Unexpected_disconect_HMI_responds_ERROR_CODE.lua:1]: Typo in script name: `disconect`.

File: test_scripts/API/ButtonSubscription/030_SubscribeButton_on_different_buttons_Unexpected_disconect.lua [hunk 15]
```diff
@@ -0,0 +1,80 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/030_SubscribeButton_on_different_buttons_Unexpected_disconect.lua:1]: Typo in script name: `disconect`.

File: test_scripts/API/ButtonSubscription/031_SubscribeButton_Unexpected_disconect_HMI_doesn_t_response_timeout.lua [hunk 16]
```diff
@@ -0,0 +1,59 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/031_SubscribeButton_Unexpected_disconect_HMI_doesn_t_response_timeout.lua:1]: Typo in script name: `disconect`.

File: test_scripts/API/ButtonSubscription/041_Subscribe_CUSTOM_BUTTON_Resumption_Unexpected_disconect.lua [hunk 17]
```diff
@@ -0,0 +1,57 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/041_Subscribe_CUSTOM_BUTTON_Resumption_Unexpected_disconect.lua:1]: Typo in script name: `disconect`.

File: test_scripts/API/ButtonSubscription/042_Unsubscribe_CUSTOM_BUTTON_Resumption_Unexpected_disconect.lua [hunk 18]
```diff
@@ -0,0 +1,52 @@
+------------------------------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/042_Unsubscribe_CUSTOM_BUTTON_Resumption_Unexpected_disconect.lua:1]: Typo in script name: `disconect`.

File: test_scripts/API/ButtonSubscription/001_Subscribe_Unsubscribe_Button_Success.lua [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+common.runner.Title("Test")
+for _, buttonName in common.spairs(common.buttons) do
+	common.runner.Title("ButtonName parameter: " .. buttonName)
+	common.runner.Step("SubscribeButton " .. buttonName, common.rpcSuccess,
+		{ appSessionId1, "SubscribeButton", buttonName })
+	common.runner.Step("On Button Press " .. buttonName, common.buttonPress, { appSessionId1, buttonName })
+	common.runner.Step("UnsubscribeButton " .. buttonName, common.rpcSuccess,
+		{ appSessionId1, "UnsubscribeButton", buttonName })
+	common.runner.Step("Check unsubscribe " .. buttonName, common.buttonPress,
+		{ appSessionId1, buttonName, common.isNotExpected })
```
Comment (dboltovskyi) [inline test_scripts/API/ButtonSubscription/001_Subscribe_Unsubscribe_Button_Success.lua:38]: Please replace all tabs by spaces => as it's already done in other scripts

**Review by dboltovskyi — CHANGES_REQUESTED** [review]

**Review by dboltovskyi — CHANGES_REQUESTED** [review]
Please also check scripts correspond to test set in Jira

**Pull request closed without merging**